Improve powershell startup time

daxian-dbw opened the pull request:
## PR Summary

With this PR, the startup time of `pwsh` gets about `24.5%` improvement on average comparing with the PS Core 6.1.0 official release. The following is an example measurement result by using `BenchmarkDotNet v0.11.2` (the detailed information about this measurement result can be found at the end):

```
# Warm startup -- with module analysis cache pre-populated
pwsh.exe -noprofile -c echo 1

           Method |     Mean |    Error |    StdDev |
----------------- |---------:|---------:|----------:|
           Pwsh61 | 540.5 ms | 9.381 ms |  8.316 ms |
            WinPS | 332.9 ms | 6.497 ms | 10.675 ms |
 pwsh62_optimized | 408.3 ms | 8.936 ms |  9.176 ms |
```

### Detailed Changes
- Avoid `SecuritySupport.IsProductBinary` and unnecessary AMSI/suspicious code scan at startup time for a regular session
   - Update `CompiledScriptBlockData.IsProductCode` to avoid unnecessary calls to `IsProductBinary`, which attempts to retrieve catalog signature of the target file.
   - Update `PerformSecurityChecks` to skip AMSI and suspicious code scan for the `.psd1` file that contains a safe `HashtableAst` only.
- Use customized `ReadOnlyBag` instead of `ImmutableHashSet` so that we can avoid loading the `System.Collections.Immutable.dll` completely.
- Replace `SHA1` with `CRC32` when generating module analysis cache file name
   - This remove the loading of `System.Security.Cryptography.Algorithms.dll` at startup
- Move `ConvertFrom-SddlString` to C# to remove the `Utility.psm1` file.
- Crossgen `Microsoft.ApplicationInsights.dll` and enable tiered compilation
   - Even pwsh with crossgen assemblies spends a lot time in jitting at the startup, about `191.6ms` comparing with `24.7ms` for Windows PowerShell.
   - Jitting `Microsoft.ApplicationInsights.dll` took about `51.6ms`.
   - By crossgen `Microsoft.ApplicationInsights.dll` and enable tiered compilation, the jitting time drops to about `98.9ms`. Detailed information can be found at the end.

**NOTE: Each commit is self-contained and the purpose of the commit is described by the commit message. So it would be easier to review per commit.**

### Compared with Windows PowerShell

The optimized pwsh is still about `75ms` slower than Windows PowerShell. The main reason is jitting time difference, even though we crossgen all the assemblies in the table below except for `pwsh.dll`.

More work will be done to further bring down the jitting time at startup.

### Reference Set Improvement

Impacting memory size loaded from disk (total): 5.9% less pages (606 / 10289 pages), 6% less memory (2.4 / 40.2 mb). The comparison of the image pages being loaded during startup before and after this PR is as follows:

```none
TotalImagesAdded   : 0
TotalImagesRemoved : 15
TotalImagesChanged : 19
TotalPageChange    : -556


   type: Remove

type   image                                       pages
----   -----                                       -----
Remove AppxSip.dll                                 21
Remove coml2.dll                                   31
Remove iertutil.dll                                42
Remove mintdh.dll                                  22
Remove msisip.dll                                  7
Remove OpcServices.dll                             53
Remove pwrshsip.dll                                11
Remove System.Collections.Immutable.dll            74
Remove System.Security.Cryptography.Algorithms.dll 38
Remove System.Security.Permissions.dll             14
Remove System.Threading.AccessControl.dll          6
Remove tdh.dll                                     15
Remove urlmon.dll                                  66
Remove wshext.dll                                  12
Remove xmllite.dll                                 13

   type: Update

type   image                                             pages
----   -----                                             -----
Update clrjit.dll                                        -4
Update combase.dll                                       1
Update coreclr.dll                                       -10
Update crypt32.dll                                       -70
Update KernelBase.dll                                    -3
Update Microsoft.PowerShell.Commands.Utility.dll         4
Update msasn1.dll                                        -4
Update msvcrt.dll                                        -1
Update ntdll.dll                                         1
Update rpcrt4.dll                                        -2
Update sechost.dll                                       -5
Update SHCore.dll                                        -10
Update shell32.dll                                       -1
Update System.Private.CoreLib.dll                        -7
Update System.Runtime.Extensions.dll                     -1
Update System.Security.Cryptography.Primitives.dll       -3
Update System.Security.Cryptography.X509Certificates.dll -2
Update windows.storage.dll                               2
Update wintrust.dll                                      -16
```

### Jitting Time Improvement

#### Windows PowerShell

Windows PowerShell Assembly Name | JitTime   msec | Num   Compilations | IL   Size | Native   Size | Foreground   msec | Multicore   JIT Background   msec | Tiered   Compilation Background   msec
-- | -- | -- | -- | -- | -- | -- | --
TOTAL | 24.7 | 83 | 1,506 | 16,149 | 24.7 | 0.0 | 0.0
Anonymously Hosted DynamicMethods Assembly | 19.3 | 44 | 0 | 12,517 | 19.3 | 0.0 | 0.0
C:\WINDOWS\Microsoft.Net\assembly\GAC_MSIL\...\System.Management.Automation.dll | 4.3 | 38 | 1,181 | 2,782 | 4.3 | 0.0 | 0.0
C:\WINDOWS\Microsoft.Net\assembly\GAC_MSIL\...\System.Core.dll | 1.1 | 1 | 325 | 850 | 1.1 | 0.0 | 0.0

#### Before crossgen'ing `Microsoft.ApplicationInsights.dll` and applying tiered compilation

PWSH   Assembly Name | JitTime   msec | Num   Compilations | IL   Size | Native   Size | Foreground   msec | Multicore   JIT Background   msec | Tiered   Compilation Background   msec
-- | -- | -- | -- | -- | -- | -- | --
TOTAL | 191.6 | 488 | 43,105 | 140,540 | 151.2 | 40.4 | 0.0
E:\PowerShell\bin\latest\Microsoft.ApplicationInsights.dll | 51.6 | 147 | 9,866 | 30,735 | 51.6 | 0.0 | 0.0
E:\PowerShell\bin\latest\System.Management.Automation.dll | 44.4 | 25 | 14,661 | 44,123 | 4.0 | 40.4 | 0.0
E:\PowerShell\bin\latest\System.Private.CoreLib.dll | 36.0 | 143 | 11,947 | 20,691 | 36.0 | 0.0 | 0.0
E:\PowerShell\bin\latest\Microsoft.PowerShell.CoreCLR.Eventing.dll | 22.1 | 80 | 0 | 23,342 | 22.1 | 0.0 | 0.0
E:\PowerShell\bin\latest\System.Collections.Concurrent.dll | 17.6 | 34 | 4,018 | 12,261 | 17.6 | 0.0 | 0.0
E:\PowerShell\bin\latest\System.Collections.dll | 9.3 | 34 | 1,317 | 3,542 | 9.3 | 0.0 | 0.0
E:\PowerShell\bin\latest\System.Linq.dll | 7.6 | 17 | 1,085 | 4,911 | 7.6 | 0.0 | 0.0
Anonymously Hosted DynamicMethods Assembly | 1.4 | 6 | 0 | 543 | 1.4 | 0.0 | 0.0
E:\PowerShell\bin\latest\Newtonsoft.Json.dll | 0.8 | 1 | 196 | 357 | 0.8 | 0.0 | 0.0
E:\PowerShell\bin\latest\pwsh.dll | 0.7 | 1 | 15 | 35 | 0.7 | 0.0 | 0.0

#### After crossgen `Microsoft.ApplicationInsights.dll` and applying tiered compilation


Updated   PWSH Assembly Name | JitTime   msec | Num   Compilations | IL   Size | Native   Size | Foreground   msec | Multicore   JIT Background   msec | Tiered   Compilation Background   msec
-- | -- | -- | -- | -- | -- | -- | --
TOTAL | 98.9 | 422 | 34,500 | 144,231 | 78.3 | 20.5 | 0.0
E:\PowerShell\bin\base\System.Private.CoreLib.dll | 26.7 | 181 | 12,442 | 36,649 | 26.7 | 0.0 | 0.0
E:\PowerShell\bin\base\System.Management.Automation.dll | 22.7 | 25 | 14,661 | 51,754 | 2.4 | 20.3 | 0.0
E:\PowerShell\bin\base\Microsoft.PowerShell.CoreCLR.Eventing.dll | 22.1 | 80 | 0 | 22,778 | 22.1 | 0.0 | 0.0
E:\PowerShell\bin\base\System.Collections.Concurrent.dll | 9.9 | 53 | 4,429 | 19,337 | 9.6 | 0.3 | 0.0
E:\PowerShell\bin\base\System.Collections.dll | 7.0 | 52 | 1,546 | 6,856 | 7.0 | 0.0 | 0.0
E:\PowerShell\bin\base\System.Linq.dll | 5.3 | 23 | 1,211 | 5,770 | 5.3 | 0.0 | 0.0
Anonymously   Hosted DynamicMethods Assembly | 4.1 | 6 | 0 | 543 | 4.1 | 0.0 | 0.0
E:\PowerShell\bin\base\pwsh.dll | 0.7 | 1 | 15 | 51 | 0.7 | 0.0 | 0.0
E:\PowerShell\bin\base\Newtonsoft.Json.dll | 0.3 | 1 | 196 | 493 | 0.3 | 0.0 | 0.0



### Startup Detailed Measurement Record

```none
// * Detailed results *
PSStartupTime.Pwsh61: Core(Runtime=Core)
Runtime = .NET Core 2.1.6 (CoreCLR 4.6.27019.06, CoreFX 4.6.27019.05), 64bit RyuJIT; GC = Concurrent Workstation
Mean = 540.5245 ms, StdErr = 2.2226 ms (0.41%); N = 14, StdDev = 8.3163 ms
Min = 528.3226 ms, Q1 = 535.3209 ms, Median = 538.9606 ms, Q3 = 544.9347 ms, Max = 559.4117 ms
IQR = 9.6138 ms, LowerFence = 520.9002 ms, UpperFence = 559.3554 ms
ConfidenceInterval = [531.1432 ms; 549.9058 ms] (CI 99.9%), Margin = 9.3813 ms (1.74% of Mean)
Skewness = 0.66, Kurtosis = 2.66, MValue = 2
-------------------- Histogram --------------------
[525.303 ms ; 539.307 ms) | @@@@@@@
[539.307 ms ; 562.431 ms) | @@@@@@@
---------------------------------------------------

PSStartupTime.WinPS: Core(Runtime=Core)
Runtime = .NET Core 2.1.6 (CoreCLR 4.6.27019.06, CoreFX 4.6.27019.05), 64bit RyuJIT; GC = Concurrent Workstation
Mean = 332.9168 ms, StdErr = 1.8044 ms (0.54%); N = 35, StdDev = 10.6749 ms
Min = 316.7778 ms, Q1 = 323.8313 ms, Median = 330.4381 ms, Q3 = 340.1557 ms, Max = 356.9562 ms
IQR = 16.3244 ms, LowerFence = 299.3447 ms, UpperFence = 364.6423 ms
ConfidenceInterval = [326.4197 ms; 339.4139 ms] (CI 99.9%), Margin = 6.4971 ms (1.95% of Mean)
Skewness = 0.47, Kurtosis = 2.28, MValue = 2.2
-------------------- Histogram --------------------
[313.922 ms ; 324.935 ms) | @@@@@@@@@@
[324.935 ms ; 332.120 ms) | @@@@@@@@@
[332.120 ms ; 341.104 ms) | @@@@@@@@@
[341.104 ms ; 348.303 ms) | @@@
[348.303 ms ; 359.812 ms) | @@@@
---------------------------------------------------

PSStartupTime.pwsh62_optimized: Core(Runtime=Core)
Runtime = .NET Core 2.1.6 (CoreCLR 4.6.27019.06, CoreFX 4.6.27019.05), 64bit RyuJIT; GC = Concurrent Workstation
Mean = 408.2847 ms, StdErr = 2.2256 ms (0.55%); N = 17, StdDev = 9.1764 ms
Min = 387.1787 ms, Q1 = 403.7134 ms, Median = 409.1181 ms, Q3 = 414.1094 ms, Max = 428.8114 ms
IQR = 10.3960 ms, LowerFence = 388.1194 ms, UpperFence = 429.7034 ms
ConfidenceInterval = [399.3489 ms; 417.2204 ms] (CI 99.9%), Margin = 8.9358 ms (2.19% of Mean)
Skewness = -0.18, Kurtosis = 3.46, MValue = 2
-------------------- Histogram --------------------
[384.056 ms ; 395.103 ms) | @
[395.103 ms ; 403.504 ms) | @@@
[403.504 ms ; 418.618 ms) | @@@@@@@@@@@@
[418.618 ms ; 431.934 ms) | @
---------------------------------------------------

// * Summary *

BenchmarkDotNet=v0.11.2, OS=Windows 10.0.17763.134 (1809/October2018Update/Redstone5)
Intel Core i7-6700 CPU 3.40GHz (Skylake), 1 CPU, 8 logical and 4 physical cores
.NET Core SDK=2.1.500
  [Host] : .NET Core 2.1.6 (CoreCLR 4.6.27019.06, CoreFX 4.6.27019.05), 64bit RyuJIT
  Core   : .NET Core 2.1.6 (CoreCLR 4.6.27019.06, CoreFX 4.6.27019.05), 64bit RyuJIT

Job=Core  Runtime=Core

           Method |     Mean |    Error |    StdDev |
----------------- |---------:|---------:|----------:|
           Pwsh61 | 540.5 ms | 9.381 ms |  8.316 ms |
            WinPS | 332.9 ms | 6.497 ms | 10.675 ms |
 pwsh62_optimized | 408.3 ms | 8.936 ms |  9.176 ms |

// * Hints *
Outliers
  PSStartupTime.Pwsh61: Core           -> 1 outlier  was  removed
  PSStartupTime.pwsh62_optimized: Core -> 1 outlier  was  detected

// * Legends *
  Mean   : Arithmetic mean of all measurements
  Error  : Half of 99.9% confidence interval
  StdDev : Standard deviation of all measurements
  1 ms   : 1 Millisecond (0.001 sec)

// ***** BenchmarkRunner: End *****
Run time: 00:00:38 (38.84 sec), executed benchmarks: 3
```


## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` to the beginning of the title and remove the prefix when the PR is ready.
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [x] User-facing [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed - Issue link:
- **Testing - New and feature**
    - [ ] Not Applicable or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
        - [x] [Add `[feature]` if the change is significant or affects feature tests](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#requesting-additional-tests-for-a-pr)

daxian-dbw committed:
Crossgen 'Microsoft.ApplicationInsights' and enable tiered compilation

daxian-dbw committed:
[Feature] Update '.gitignore' to ignore .vscode folders from the sub directories

daxian-dbw committed:
Avoid 'SecuritySupport.IsProductBinary' and unnecessary AMSI/suspicious code scan at startup time for a regular session

daxian-dbw committed:
Use customized 'ReadOnlyHashSet' instead of 'ImmutableHashSet'

daxian-dbw committed:
Replace SHA1 with CRC32 when generating module analysis cache file name

daxian-dbw committed:
Move 'ConvertFrom-SddlString' to C# to avoid Utility.psm1 completely

daxian-dbw committed:
Update the packaging/signing metadata due to removal of utility.psm1

iSazonov commented:
> More work will be done to further bring down the jitting time at startup.

Do you continue in the PR?

> After crossgen Microsoft.ApplicationInsights.dll and applying tiered compilation

I don't see Microsoft.ApplicationInsights.dll in the table - no jitting?

I searched "Insights" in our code base and have questions:
1. Do we really need types from the dll in CorePsTypeCatalog.cs? We could remove 239 lines.
2. The dll is used for telemetry. Do we really need load the dll at startup? Seems we could delay the telemetry
3. The telemetry is a static class. We have other static classes. I guess some static classes is initialized at startup (I guess they is initialized at first use time) - does this slow down startup? Make sense to analyze this and try to delay static classes initializations?

If `ReadOnlyHashSet` was rewrited I think we could look other classes used in PowerShell Core engine. Maybe it won't speed up the startup but increase performance.

```
Outliers
  PSStartupTime.Pwsh61: Core           -> 2 outliers were removed
  PSStartupTime.WinPS: Core            -> 4 outliers were removed
  PSStartupTime.pwsh62_optimized: Core -> 1 outlier  was  removed
```
having `Outliers` it is not good in the test - seems interference from other processes.

> Intel Core i7-6700 CPU 3.40GHz (Skylake), 1 CPU, 8 logical and 4 physical cores

I think that an ordinary user computer is 2 physical cores and it would be correct to limit the test to two physical cores. (This may be related to tied compilation)

iSazonov commented:
I restarted CI-MacOs.

iSazonov commented on src/System.Management.Automation/utils/PsUtils.cs:
```diff
@@ -624,6 +624,63 @@ internal static object[] Base64ToArgsConverter(string base64)
+        private const uint polynomial = 0x1EDC6F41;
+        private static uint[] table;
+
+        static CRC32Hash()
```
Could we use pre-defined table and exclude this calculations at startup?

daxian-dbw replied:
Well, the calculated table is the same for the given const `polynomial`, so I guess this could be an option.
@TravisEz13 can you please weigh in?

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/ConvertFrom-SddlString.cs:
```diff
@@ -0,0 +1,261 @@
+    /// </summary>
+    [Cmdlet(VerbsData.ConvertFrom, "SddlString", HelpUri = "https://go.microsoft.com/fwlink/?LinkId=623636", RemotingCapability = RemotingCapability.None)]
+    [OutputType(typeof(SecurityDescriptorInfo))]
+    public sealed class ConvertFromSddlStringCommand : PSCmdlet
```
I'd move the commit in separate PR  - previously we had PR #3936 to make this conversion but I had many questions and that PR was not merged.

iSazonov replied:
@daxian-dbw I agree. Please open tracking issue after the PR will merged.

daxian-dbw replied:
Issue #8365 was opened for this.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/ConvertFrom-SddlString.cs:
```diff
@@ -0,0 +1,261 @@
+            {
+                foreach (string memberName in Enum.GetNames(accessRightType))
+                {
+                    int memberValue = (int) Enum.Parse(accessRightType, memberName);
```
Please remove extra space after `(int)`.

daxian-dbw replied:
Fixed.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/UtilityCommon.cs:
```diff
@@ -59,35 +59,6 @@ public enum TextEncodingType
-    /// <summary>
-    /// Utility class to contain resources for the Microsoft.PowerShell.Utility module.
-    /// </summary>
-    public static class UtilityResources
```
Seems we can not remove public APIs. We can use `obsolete` attribute.

daxian-dbw replied:
This type was there only because we used to have script functions that use the built-in resources files. This is theoretically a breaking change, but practically I don't think it will affect anyone :)

SteveL-MSFT replied:
A [search on GitHub](https://github.com/search?l=PowerShell&q=utilityresources&type=Code) returned a few instances of scripts using this public class.  Seems like there isn't a good reason to break them.

iSazonov replied:
We could move obsolete APIs (we have others) to separate dll and perhaps to separate compatibility package.

daxian-dbw replied:
For obsoleted types, you may be able to do that by type forwarding. For obsoleted type members, I don't think there is a way to do that. This should be a separate discussion.

iSazonov commented on src/Modules/Unix/Microsoft.PowerShell.Utility/Microsoft.PowerShell.Utility.psd1:
```diff
@@ -1,32 +1,34 @@
-AliasesToExport= "fhx"
-NestedModules="Microsoft.PowerShell.Commands.Utility.dll","Microsoft.PowerShell.Utility.psm1"
+PowerShellVersion = "3.0"
+CmdletsToExport = @(
```
When I see the list, I always think about alphabetical sorting.

daxian-dbw replied:
It is sorted alphabetically, on the noun part of the command name, as you can see `Alias`, `Clixml`, `Command`, `Csv` ... I think related commands should be grouped together. This is not perfect though, because for `Format-*` commands, they should be grouped based on the verb part ...

daxian-dbw replied:
Are you suggesting list all commands as a 1-column list like in `management.psd1`? There are 57 cmdlets in `management.psd1` but `108` in `utility.psd1`. The 1-column list would be too long.

iSazonov replied:
The 1-columns sort looks good to me.

daxian-dbw replied:
I think it's probably better to group related cmdlets together in the same row, so the list will be like this:
```
'Get-Alias', 'Set-Alias', 'Remove-Alias',
'Import-Csv', 'Export-Csv',
...
```
But I will deffer this change to a separate follow-up PR.

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Utility/ConvertFrom-SddlString.ps1:
```diff
@@ -0,0 +1,50 @@
+        $result = $sddl | ConvertFrom-SddlString @arguments
+        foreach ($property in $expectedProperties)
+        {
+            $result.$property | Should Not Be $null
```
`Should -Not -Be $null`

daxian-dbw replied:
Good catch.

iSazonov commented on src/System.Management.Automation/engine/InitialSessionState.cs:
```diff
@@ -2382,9 +2382,6 @@ internal Exception BindRunspace(Runspace initializedRunspace, PSTraceSource runs
 
-            // Set the SessionStateDrive here since we have all the provider information at this point
-            SetSessionStateDrive(initializedRunspace.ExecutionContext, true);
-
```
How the changes relate to `Microsoft.ApplicationInsights`?

daxian-dbw replied:
This is a minor optimization in CPU time by removing redundant code. `BindRunspace` is called only from `DoOpenHelper()`, after `new AutomationEngine(Host, InitialSessionState)`. `SetSessionStateDrive(context, true)` has already been called in the constructor of `AutomationEngine`.

iSazonov commented on src/powershell-win-core/powershell-win-core.csproj:
```diff
@@ -4,6 +4,7 @@
     <Description>PowerShell Core on Windows top-level project</Description>
     <AssemblyName>pwsh</AssemblyName>
     <OutputType>Exe</OutputType>
+    <TieredCompilation>true</TieredCompilation>
```
I'd add this to our tool/test projects too.

daxian-dbw replied:
I can do that.

iSazonov commented on .gitignore:
```diff
@@ -24,6 +24,9 @@ dotnet-uninstall-debian-packages.sh
 .vs/
 
+# VS Code directory
+/**/.vscode/
```
In home folder we have `.vscode` that should remain under git control.

daxian-dbw replied:
This is to ignore `.vscode` folders in the sub directories. Changes to the one in the root directory is not ignored.

lzybkr commented on src/System.Management.Automation/engine/ExperimentalFeature/ExperimentalFeature.cs:
```diff
@@ -339,4 +339,198 @@ private ExperimentAction EffectiveAction
+    /// A readonly Hashset
+    /// </summary>
+    public sealed class ReadOnlyHashSet<T> : ISet<T>
+    {
```
Public?

daxian-dbw replied:
Yes, `$EnabledExperimentalFeatures` is exposed to the user and it refers to `ExperimentalFeature.EnabledExperimentalFeatures`, which is a static object of this type.

daxian-dbw replied:
It turns out we don't need this type to be public for users to use `$EnabledExperimentalFeatures` the way they want. So I agree to change it to internal.

lzybkr replied:
It feels like you're just exposing an internal implementation detail instead of consciously deciding what api users need.

daxian-dbw replied:
You are right. Most of the members I exposed here should be exposed at all. I will revisit this.

daxian-dbw replied:
I changed most of the method to be explicit implementations of the interface methods of `ISet<T>`, but that doesn't help. `Get-Member` and tab completion still show those interface methods on `$EnabledExperimentalFeatures`, and even worse, call those methods in PowerShell fails with confusing error message, for example, "Cannot find an overload of 'Add' that takes 1 argument". I feel it's better to expose those interface methods as public ...

> It feels like you're just exposing an internal implementation detail instead of consciously deciding what api users need.

Now I'm not sure what you mean exactly. The purpose to have `$EnabledExperimentalFeatures` is to allow the user to conditionally expose cmdlets in module manifest. Say a module defines an experimental feature, which involves a new cmdlet that should be exposed only if the experimental feature is enabled. Also, it allows users to check what experimental features are enabled, so different code paths can be taken in script. Can you please elaborate a bit more on your concerns here?

iSazonov replied:
Should we have `$EnabledExperimentalFeatures` only in preview versions? If we remove it from Release we could remove `ReadOnlyHashSet` from release too.

daxian-dbw replied:
I got your point now. Thanks!
I thought about only exposing a method for checking enabled features, but `$ExperimentalFeatures` will mainly be used in module manifest files for deciding what set of cmdlets/functions/aliases to expose from a module depending on whether or not an experimental feature defined in the module is enabled. In a module manifest file, you cannot call methods, and thus you have to check in this way: `if ($ExperimentalFeatures -contains "MyExpFeature") { ... } else { ... }`. So the object referred by `$ExperimentalFeatures` at least has to implement `IEnumerable`. I make it implement `ISet` because I want to make this customized type as general as possible, so that it might be reused in other places in the code base in future.

lzybkr replied:
`-contains` doesn't do anything special for `ISet`. You could expose `IEnumerable` and it will work with `-contains`.

Another option is to make this type implement `IDynamicMetaObjectProvider` so you can write:

```powershell
if ($ExperimentalFeatures.MyFeature) { ... }
```

daxian-dbw replied:
Property reference is not allowed in module manifest file.
I will make the type only implements `IEnumerable` for now, and move it to `S.M.A.Internal` namespace. If there is a need for reusing it in future as a set, it can always be updated to implement `ISet`.

lzybkr commented on src/System.Management.Automation/engine/ExperimentalFeature/ExperimentalFeature.cs:
```diff
@@ -119,11 +119,11 @@ static ExperimentalFeature()
+        /// return an ReadOnlyHashSet of the valid enabled feature names.
         /// </summary>
-        private static ImmutableHashSet<string> ProcessEnabledFeatures(string[] enabledFeatures)
+        private static ReadOnlyHashSet<string> ProcessEnabledFeatures(string[] enabledFeatures)
```
>ReadOnlyHashSet [](start = 23, length = 15)

Why not `ReadOnlyDictionary`? You could avoid adding a new type that way.

daxian-dbw replied:
The `HashSet` is exposed to user via `$EnabledExperimentalFeatures`. I think it's weird to have a `ReadOnlyDictionary` object for this variable.

lzybkr replied:
I think it's weird to have `System.Management.Automation.ReadOnlyHashSet`.

How about making it a `HashSet` that is just a clone, or cloned on every access?

daxian-dbw replied:
How about changing the type to `internal`? Since the type has public members, changing it to internal doesn't affect users from using `$EnabledExperimentalFeatures` the way they want (access `Count`, call `Contains(string)`).

daxian-dbw commented on src/System.Management.Automation/engine/runtime/CompiledScriptBlock.cs:
```diff
@@ -193,15 +192,28 @@ private void ReallyCompile(bool optimize)
             var scriptBlockAst = Ast as ScriptBlockAst;
-            if (scriptBlockAst == null)
+            if (scriptBlockAst == null || _isProductCode == true)
             {
```
@PaulHigin @TravisEz13 can you please weigh in here?

My main purpose here is to avoid the false positive result from the suspicious scan on `Utility.psd1`, which declares `Add-Type` in it. Due to this false positive scan result, powershell will attempt to log the script block when evaluating `Utility.psd1`, which will trigger the call to `IsProductSignature` that triggers the expensive catalog signature invocations.

iSazonov replied:
In fact, we come from one point -  `ScriptBlock.CreateDelayParsedScriptBlock("string", isProductCode: true)`.
I don't review in depth but I don't found the pattern "isProductCode: false". Perhaps we could implement separate code path.

daxian-dbw replied:
Cool, I will remove the check for `_isProductCode == true` here.
Just to bring it up that `IsProductCode` is used in `LogScriptBlockCreation`  to decide whether we need to log the script text. So an attacker can still change this property and make the malicious script not logged in the event log.

daxian-dbw replied:
The check is removed.

lzybkr commented on src/System.Management.Automation/engine/Modules/AnalysisCache.cs:
```diff
@@ -1071,45 +1070,37 @@ static AnalysisCacheData()
+                // Use CRC32 because it's faster.
+                // It's very unlikely to get collision from hashing the combinations of enabled features names.
+                hashString = CRC32Hash.ComputeHash(allNames);
+                cacheFileName = string.Format(CultureInfo.InvariantCulture, "{0}-{1}", cacheFileName, hashString);
```
Why not `GetHashCode`, or if that's not stable across runs, the a simple non-allocating hash function.

daxian-dbw replied:
`GetHashCode` in .NET Core 2.1 returns different values from different processes for the same string. That's a by design change (Marvin algorithm) in .NET Core to prevent DOS attack.

lzybkr replied:
But you can have a simpler stable hash function, e.g. this [one](https://github.com/PowerShell/PowerShell/blob/b0daa1e724cbba5e00f118001f1105bb44d91952/src/System.Management.Automation/engine/runtime/CompiledScriptBlock.cs#L1804)

daxian-dbw replied:
Interesting, I'm more than happy to replace CRC32 with a simpler stable hash function. Will take a look.

iSazonov replied:
Seems I know how to make the simple function faster. :-) https://github.com/dotnet/corefx/pull/32949#discussion_r236468915

iSazonov replied:
Update: https://source.dot.net/#System.Private.CoreLib/shared/System/Globalization/TextInfo.cs,609

daxian-dbw replied:
Thanks @iSazonov. Wait on the feedback from other folks. Once we land on the simple hash function, I can make the change in a separate follow-up PR.

daxian-dbw replied:
I opened #8364 to track the follow-up work for myself. I will fix it in a separate PR.

SteveL-MSFT replied:
Although the hash function is called is one-time for the module analysis, it's called at startup every time.  The [MurmurHash3](https://github.com/aappleby/smhasher) might be a better choice for speed and uniqueness (avoiding collisions), but we can defer that to outside this PR if we open an issue.

lzybkr replied:
I think you should prefer the simplest possible code for this use case. The actual hash function isn't too important, it's the side effects (like loading an assembly, memory allocation) that I think you want to avoid.

daxian-dbw replied:
Thanks for all the discussion here. I will take all input into account when working on the hash function PR.

lzybkr commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/Import-LocalizedData.cs:
```diff
@@ -167,6 +167,7 @@ protected override void ProcessRecord()
                 try
                 {
+                    scriptBlock.ScriptBlockData.ForceMaliciousCodeScan = true;
                     result = scriptBlock.InvokeReturnAsIs();
```
Assuming you don't squash this PR, this change seems to be in the wrong commit.

daxian-dbw replied:
Why? I check for `ForceMaliciousCodeScan` before doing AMSI/suspicious code scan when the script block is from a `.psd1` file. The reason to have this new property in `CompiledScriptBlockData` is to avoid scanning `.psd1` file unless it's being evaluated from `Import-LocalizedData`.

lzybkr replied:
Hmm, CodeFlow might be lying to me, it shows as a diff b/w iterations 3 and 4, but it also shows the diff in iteration 1.

daxian-dbw commented on src/System.Management.Automation/engine/runtime/CompiledScriptBlock.cs:
```diff
@@ -49,7 +50,10 @@ internal CompiledScriptBlockData(IParameterMetadataProvider ast, bool isFilter)
+            if (isProductCode)
+            {
+                _isProductCode = true;
+            }
```
Good point. Maybe it's safe to just use `_isProductCode = isProductCode` here, as it's explicitly passed in.

daxian-dbw commented on src/System.Management.Automation/engine/runtime/CompiledScriptBlock.cs:
```diff
@@ -193,15 +192,28 @@ private void ReallyCompile(bool optimize)
+
+            if (scriptFile != null &&
+                StringLiterals.PowerShellDataFileExtension.Equals(
+                    Path.GetExtension(scriptFile), StringComparison.OrdinalIgnoreCase)
```
Even more concise. I like it.

daxian-dbw commented on src/System.Management.Automation/engine/runtime/CompiledScriptBlock.cs:
```diff
@@ -193,15 +192,28 @@ private void ReallyCompile(bool optimize)
+            if (scriptFile != null &&
+                StringLiterals.PowerShellDataFileExtension.Equals(
+                    Path.GetExtension(scriptFile), StringComparison.OrdinalIgnoreCase)
+                && !ForceMaliciousCodeScan)
```
Currently `ForceMaliciousCodeScan` is almost always `false` unless the script block is from `Import-LocalizedData`, so I would keep it as is for now.

daxian-dbw replied:
If two of you believe so, I will make the change :) The current order is just micro-optimization that won't affect the startup time.

daxian-dbw replied:
Fixed.

daxian-dbw commented on src/System.Management.Automation/engine/Modules/AnalysisCache.cs:
```diff
@@ -1071,45 +1070,37 @@ static AnalysisCacheData()
-
-                string hashString = BitConverter.ToString(hashBytes, startIndex: 0, length: 4).Replace("-", string.Empty);
-                cacheFileName = string.Format(CultureInfo.InvariantCulture, "{0}-{1}", cacheFileName, hashString);
+            string hashString = CRC32Hash.ComputeHash(Utils.DefaultPowerShellAppBase);
```
Yes, but it should be fine. The `SHA1` hashing on `$pshome` was introduced recently, there won't be too much impact.

daxian-dbw commented on src/Microsoft.PowerShell.Commands.Utility/Microsoft.PowerShell.Commands.Utility.csproj:
```diff
@@ -67,6 +67,7 @@
 
   <ItemGroup>
     <PackageReference Include="Microsoft.CodeAnalysis.CSharp" Version="2.10.0" />
+    <PackageReference Include="System.Threading.AccessControl" Version="4.5.0" />
```
No, I just moved the package reference from `SDK` to `Uitlity`.

daxian-dbw commented:
@iSazonov Reply to your comment here:

> > More work will be done to further bring down the jitting time at startup.
>
> Do you continue in the PR?

No, that will be in a new PR. The planned work are as follows:
- Generate PDB for crossgen'ed assemblies that can be used by PerfView/WPA, so that I can do further stack analysis on CPU time.
- Crossgen'ed assemblies still leave a high cost of jitting time comparing with ngen. It's probably by design as ngen is targeting a specific CPU while crossgen is targeting a general x64 architecture, but I need to contact .NET Core team to get better understanding and see if there are other options (profile-guided-opt?)
- Ad-hoc code changes to avoid unnecessary work at startup time.

> I don't see Microsoft.ApplicationInsights.dll in the table - no jitting?

After the change, no IL code needs to be jitted for this dll at startup time.

> Do we really need types from the dll in CorePsTypeCatalog.cs? We could remove 239 lines.

Yes we do. Type conversion depends on the type catalog. Not all .NET Core assemblies shipped with PS Core are loaded at startup. Today, user can use a type even it's from an assembly that is not loaded yet, because powershell knows which assembly the type is from by looking in the type catalog. Without it, the user needs to know which assembly the type is in, and then load the assembly first before using the type. The type catalog is also used in tab completion, so that powershell can tab complete types that are not loaded yet.

> The dll is used for telemetry. Do we really need load the dll at startup? Seems we could delay the telemetry

ApplicationInsight telemetry is done in console.dll at startup, as long as the user hasn't opted out.

> The telemetry is a static class. We have other static classes. I guess some static classes is initialized at startup (I guess they is initialized at first use time) - does this slow down startup? Make sense to analyze this and try to delay static classes initializations?

The work that is done in this PR are based on the analysis of Reference Set and Jitting time. Some ad-hoc analysis on the static members might be useful, but I think our efforts should be guided by the measurement data, so our time can be spent on things that matters most.

> If ReadOnlyHashSet was rewrited I think we could look other classes used in PowerShell Core engine. Maybe it won't speed up the startup but increase performance.

The reason to change `ImmutableHashSet` to `ReadOnlyHashSet` is because this only use of `ImmutableHashSet` causes `System.Collections.Immutable.dll` to be loaded at startup, which brings in 74 pages loaded from disk. This shows up as part of the reference set analysis. It might be useful to look at other classes, but again, I think our efforts should be guided by the measurement data.

> having Outliers it is not good in the test - seems interference from other processes.

There is no interference from other pwsh processes when running the tests. `BenchmarkDotNet` starts each process many times, and sometimes there are outliers.

> I think that an ordinary user computer is 2 physical cores and it would be correct to limit the test to two physical cores. (This may be related to tied compilation)

The baseline is the same, and here we care most about the percentage of improvement instead of the actual startup time.

SteveL-MSFT commented:
The macOS failure appears to be a test code issue:

```output
2018-11-27T12:24:13.7076000Z Description: Basic os_log tests on MacOS/
2018-11-27T12:24:13.7079830Z Name:        Verifies scriptblock logging
2018-11-27T12:24:13.7084250Z message:
2018-11-27T12:24:13.7088810Z did not recieve at least 18 records but 47 instead.
2018-11-27T12:24:13.7094210Z stack-trace:
2018-11-27T12:24:13.7097880Z at <ScriptBlock>, /Users/vsts/agent/2.142.1/work/1/s/test/tools/Modules/PSSysLog/PSSysLog.psm1: line 904
2018-11-27T12:24:13.7098430Z 904:             throw "did not recieve at least $MinimumCount records but $($log.Count) instead."
```

(note that `receive` is spelled wrong).  According to the text, the test expects >= 18, but 47 is >= 18, so the validation must be wrong.

daxian-dbw commented on src/System.Management.Automation/engine/ExperimentalFeature/ExperimentalFeature.cs:
```diff
@@ -339,4 +339,198 @@ private ExperimentAction EffectiveAction
     }
+
+    /// <summary>
+    /// A readonly Hashset
```
Fixed.

daxian-dbw committed:
[Feature] fix macOS build + 1st attempt to address review comments

daxian-dbw commented:
For the logging test that was failing in previous macOS CI, it's caused by removing `Utility.psm1`. The test script `$testScriptPath` has `Write-Verbose` in it, so when running `& $powershell -NoProfile -SettingsFile $configFile -Command $testScriptPath`, it triggers the auto-loading of `Utility` module. Before this PR, that results in 19 event logs. With this PR, two events that record the starting and completion of `Utility.psm1` execution are gone, so the total number of event logs becomes 17. I have update the test code, also update `PSSysLog.psm1` to correct the wrong message as @SteveL-MSFT pointed out above.

iSazonov commented:
> Yes we do. Type conversion depends on the type catalog. Not all .NET Core assemblies shipped with PS Core are loaded at startup. 

But the dll is loaded at startup. Only "Type conversion depends on the type catalog" prevents us from removing the dll from TypeCatalog?

> ApplicationInsight telemetry is done in console.dll at startup, as long as the user hasn't opted out.

I understand. My request is - Does the telemetry delay startup or no? If yes we could postpone the telemetry call and loading the dll.

> There is no interference from other pwsh processes when running the tests. BenchmarkDotNet starts each process many times, and sometimes there are outliers.

I think these outlies is important to investigate in the startup context. Outlier is _unexpected_ delay. Why we see unexpected delay in the tests? 
It is not for the PR but we shouldn't ignore this.

daxian-dbw commented:
> But the dll is loaded at startup. Only "Type conversion depends on the type catalog" prevents us from removing the dll from TypeCatalog?

Not sure what you mean by "But the dll is loaded at startup". Most of the assemblies included in the type catalog file are not loaded at startup.

Even for the dlls that are loaded at startup, type resolution would be much faster for a type from those dlls because we can avoid the effort to search types from all loaded assemblies to find the one.

> I understand. My request is - Does the telemetry delay startup or no? If yes we could postpone the telemetry call and loading the dll.

The telemetry definitely play a role in the startup time, but I don't know how much after crossgen'ing the dll because I'm not able to analyze the stack info in PerfView/WPA with crossgen'ed assemblies yet.

PaulHigin commented on src/System.Management.Automation/security/SecuritySupport.cs:
```diff
@@ -1595,6 +1595,13 @@ internal static AmsiNativeMethods.AMSI_RESULT WinScanContent(string content, str
                         }
                     }
 
+                    if (warmUp)
```
Why is this better than Init()?

daxian-dbw replied:
As explained above, the warm-up will create the `s_amsiSession` as well. And it prevents a race condition, even though the chance to hit that race condition is low.

daxian-dbw commented on src/System.Management.Automation/engine/InitialSessionState.cs:
```diff
@@ -41,15 +41,13 @@ internal static void Init()
-                // on a file we are very likely to read anyway.
-                var pshome = Utils.DefaultPowerShellAppBase;
-                var unused = SecuritySupport.IsProductBinary(Path.Combine(pshome, "Modules", "Microsoft.PowerShell.Utility", "Microsoft.PowerShell.Utility.psm1"));
+                AmsiUtils.WinScanContent(content: string.Empty, sourceMetadata: string.Empty, warmUp: true);
```
`Init()` only initialize the AMSI, but not creating a session. I think it's better to create the session as well. Also, if we only call `Init()` here, there is a slight chance of race condition in case the pipeline threads gets to `WinScanContent` and start to run `init` in a lock while the `init` in a background thread hasn't finished yet.

PaulHigin commented on src/System.Management.Automation/engine/hostifaces/LocalConnection.cs:
```diff
@@ -678,30 +678,27 @@ private void DoOpenHelper()
             s_runspaceInitTracer.WriteLine("runspace opened successfully");
 
             // Now do initial state configuration that requires an active runspace
-            if (InitialSessionState != null)
```
Do we know that InitialSessionState is always non-null now?  Should we add an assert here?

daxian-dbw replied:
Yes, it's always not null now. There is an assert at the beginning of `DoOpenHelper()`.

daxian-dbw commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/ConvertFrom-SddlString.cs:
```diff
@@ -0,0 +1,261 @@
+    /// <summary>
+    /// Representation of a security descriptor.
+    /// </summary>
+    public class SecurityDescriptorInfo
```
Fixed.

daxian-dbw committed:
[Feature] 2nd attempt to address review comments
Major change is about how we decide to skip AMSI/malicious code scan.

A script block may comes from a .psd1 file but still has arbitrary code in it.
An attacker can parse an .psd1 file to get an Ast and call Ast.GetScriptBlock()
to create a script block. That script block comes from a .psd1 file that may
contain arbitrary content.

The skip-scan condition is changed to:
- the script block should come from a .psd1 file
- the script block is essentially a HashtableAst
- the HashtableAst is safe.

daxian-dbw commented:
@lzybkr @PaulHigin @TravisEz13 Can you please look at the new changes to `PerformSecurityChecks()` in `CompiledScriptBlock.cs` again?

Major change is on how we decide to skip AMSI/malicious code scan. The previous way to check for skipping `.psd1` files causes security vulnerabilities.

An attacker can put arbitrary content in a `.psd1` file, can then call `[Parser]::ParseFile()` to get an AST from the `.psd1` file. Then the attacker can create a script block from the Ast by calling `$myAst.GetScriptBlock()`. That script block then will be considered from a `.psd1` file and skip the code scan.

The skip-scan condition is changed to:
- the script block should come from a .psd1 file
- the script block is essentially a HashtableAst
- the HashtableAst is safe, by calling `IsSafeValueVisitor.IsAstSafe(hashtableAst, GetSafeValueVisitor.SafeValueContext.Default)`.

The API `IsSafeValueVisitor.IsAstSafe` has been used in `GetSafeValueVisitor.GetSafeValue`, which is used in module script analysis code and the public API `Ast.SafeGetValue()`.

iSazonov commented:
We have many comments in the PR that is a problrm for GitHub web interface. If security concerns have been fixed we could merge and continue with hash function in follow PR.

PaulHigin commented on src/System.Management.Automation/engine/runtime/CompiledScriptBlock.cs:
```diff
@@ -228,6 +226,40 @@ private void PerformSecurityChecks()
                 HasSuspiciousContent = true;
             }
+
+            // A local function to check if the ScriptBlockAst is essentially a safe HashtableAst.
```
Is it 'essentially' a hash table or is it in fact a hash table?  Is there a difference?

PaulHigin replied:
I think a better comment is "Can skip code scan if the file is a psd1 file representing a safe hash table.", or to that effect.

daxian-dbw replied:
I will update the comment. Thanks!

PaulHigin commented on src/System.Management.Automation/engine/runtime/CompiledScriptBlock.cs:
```diff
@@ -200,10 +199,9 @@ private void PerformSecurityChecks()
 
             if (scriptFile != null &&
                 scriptFile.EndsWith(StringLiterals.PowerShellDataFileExtension, StringComparison.OrdinalIgnoreCase)
```
It feels cleaner if all checks of the file is performed in CanSkipCodeScan ...
CanSkipCodeScan(scriptFile))
... put scriptFile null check and name extension check in local function.

daxian-dbw replied:
The reason I'm organizing it this way is to avoid calling the method in probably most cases.
In most cases, `scirptFile` will be null, and I would like to avoid the overhead of a method call in such cases. Again, it's a micro-optimization, but given that the method is made be a local function (deliberately 😄), I would like to keep the current organization.

PaulHigin reviewed (approved):
LTGM.  Just minor comments.

daxian-dbw committed:
Update comments based on Paul's feedback

iSazonov commented:
@daxian-dbw  Seems you could merge.

daxian-dbw commented:
@lzybkr Would you like to take another look before I merge?

lzybkr commented on src/System.Management.Automation/engine/Utils.cs:
```diff
@@ -1927,4 +1913,50 @@ internal T Pop()
+    /// A readonly Hashset.
+    /// </summary>
+    internal sealed class ReadOnlyHashSet<T> : IEnumerable
+    {
```
You might consider using a more specialized name since this is no longer an `ISet`.

daxian-dbw replied:
I renamed the type to be `ReadOnlyBag`. That's the most appropriate name I could come up with. `ConcurrentBag` doesn't implement `ISet`, so I just borrowed `Bag` from it.

lzybkr commented on src/System.Management.Automation/engine/runtime/CompiledScriptBlock.cs:
```diff
@@ -221,6 +226,40 @@ private void PerformSecurityChecks()
             }
+
+            // We can skip code scan if the ScriptBlockAst is from a .psd1 file and it's in fact a safe HashtableAst.
+            bool CanSkipCodeScan()
```
The comment and code do not match - it is possible to call this method where the filename is not a psd1 file.

This code is also not future proof - it would be easy to miss the need for adding code here when the ast is enhanced. I don't have a simple suggestion on how to fix though.

What you really want is structural equivalence to an oracle, ignoring the children of the hashtable, then calling `IsAstSafe`, but I can't think of a fast way to implement that.

daxian-dbw replied:
Updated the local function name and comment to make them match.

I agree the code is not future proof ... I cannot think of a better solution. I added a `NOTE` comment in the local function to call it out.

daxian-dbw committed:
[Feature] Address additional comments from Jason

daxian-dbw commented:
Attempt to address comments + rebased to resolve the conflict in `Microsoft.PowerShell.Commands.Utility.csproj`.
@lzybkr Can you please take another look? Thanks!

daxian-dbw commented:
The failed test is most likely caused by #8346. It's a feature-level test, but the commits in that PR don't have the `[feature]` tag, so the feature tests didn't run for that PR CI.

iSazonov commented:
Did we lose a bit of performance with latest commits?

daxian-dbw commented:
Can you elaborate a bit on the `unicode simple case folding`? I'm not sure what that is.

vexx32 commented:
@daxian-dbw See here for the discussion on that: https://github.com/PowerShell/PowerShell/issues/8120

agonzalezm commented:
I want to use Convertfrom-sddlstring in c#  ? what is the equivalent method  in c# and what library?
